### PR TITLE
feat(STONEINTG-843): add consoleURL plr to github checkrun 

### PR DESCRIPTION
**Pull request by kasemAlem**
We added new field to github  endpoint CheckRun and CommitStatus on the creation : 

-  CheckRun           ->        DetailsURL
-  CommitStatus   ->        TargetURL

For more details see  [STONEINTEG-843](https://issues.redhat.com/browse/STONEINTG-843) 

## Maintainers will complete the following section

- [ ] Commit messages are descriptive enough ([hints](https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/))
- [ ] Code coverage from testing does not decrease and new code is covered ([check the PR coverage on codecov](https://app.codecov.io/gh/redhat-appstudio/integration-service/pulls))
- [ ] [Controllers diagrams](https://github.com/redhat-appstudio/integration-service/tree/main/docs) are updated when PR changes controllers code  (if applicable)

### CODE REVIEW

File: git/github/github.go [hunk 1]
```diff
@@ -52,6 +53,7 @@ type CommitStatusAdapter struct {
 	State       string
 	Description string
 	Context     string
+	TargetURL   string
```
Comment (dirgim) [inline git/github/github.go:56]: I may be missing something, but I can't see where this value is set to the correct console URL in the current code?
I would expect the [FormatPipelineURL](https://github.com/redhat-appstudio/integration-service/blob/c25c29fb82da14a96095e3aeba6f0d17ee062f6c/status/format.go#L214) to be called somewhere in order to generate the URL to be set for this.

File: git/github/github.go [hunk 2]
```diff
@@ -36,6 +36,7 @@ type CheckRunAdapter struct {
 	Name           string
 	SHA            string
 	ExternalID     string
+	DetailsURL     string
```
Comment (dirgim) [inline git/github/github.go:39]: Same as below,  I think this value should be set to a formatted pipelineURL somewhere in the code.

**Comment by dirgim** [comment]
I think that the commit (and also PR) title shouldn't be marked as `fix`, but as `feat` since this is an implementation of a regular story. It should also contain the reference to the [STONEINTG-843](https://issues.redhat.com/browse/STONEINTG-843) story instead of `stoneintg`.

Also, the title is currently too long and is failing the [gitlint check](https://github.com/redhat-appstudio/integration-service/actions/runs/8465245933/job/23193433190?pr=651).

File: git/github/github.go (hunk 1 above)
Comment (kasemAlem) [reply git/github/github.go:56]: I've added an assignment to update TestPipelineRunName  [here](https://github.com/kasemAlem/integration-service/blob/32590e8a4646c8800a029fffb1125763b3a6f939/status/reporter_github.go#L181)

File: git/github/github.go (hunk 2 above)
Comment (kasemAlem) [reply git/github/github.go:39]: I've added an assignment with console URL link with  TestPipelineRunName  [here](https://github.com/kasemAlem/integration-service/blob/32590e8a4646c8800a029fffb1125763b3a6f939/status/reporter_github.go#L181)

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (dirgim) [inline .vscode/launch.json:1]: I think this file shouldn't be included in this commit.

File: status/reporter_github.go [hunk 4]
```diff
@@ -314,13 +317,16 @@ func (csu *CommitStatusUpdater) createCommitStatusAdapterForSnapshot(report Test
 		return nil, fmt.Errorf("unknown status %s for integrationTestScenario %s and snapshot %s/%s", report.Status, report.ScenarioName, snapshot.Namespace, snapshot.Name)
 	}
 
+	targetURL := FormatPipelineURL(report.TestPipelineRunName, snapshot.Namespace, *csu.logger)
```
Comment (hongweiliu17) [inline status/reporter_github.go:320]: I am thinking if we should also handle the scenario here when the `TestPipelineRunName` has not been set for CheckRun and ComitStatus? Otherwise it will produce a strange targetUrl and detailsUrl, such as for inProgres status.

Comment (kasemAlem) [reply status/reporter_github.go:320]: good point @hongweiliu17 , 
I will add a check on both assignments and print a proper message when **TestPipelineRunName** is not set yet .

Comment (kasemAlem) [reply status/reporter_github.go]: so we check if report.Status is in progress then we update the CheckRun  and CommitStatus with the Console ITS PLR URL , else we set the **detailsURL** and **targetURL** to **""**

Comment (dirgim) [reply status/reporter_github.go]: Would checking if the `report.TestPipelineRunName` is set to anything before generating a pipelineURL do the job?

Comment (kasemAlem) [reply status/reporter_github.go]: you mean like the check I've added before [detailsURL](https://github.com/kasemAlem/integration-service/blob/367141fd63a0c00ae98d69917de9c94b1fac5731/status/reporter_github.go#L182-L187)   and before the [targetURL](https://github.com/kasemAlem/integration-service/blob/367141fd63a0c00ae98d69917de9c94b1fac5731/status/reporter_github.go#L327-L332) ?

File: git/github/github.go [hunk 5]
```diff
@@ -232,6 +234,7 @@ func (c *Client) CreateCheckRun(ctx context.Context, cra *CheckRunAdapter) (*int
 		Name:       cra.Name,
 		HeadSHA:    cra.SHA,
 		ExternalID: &cra.ExternalID,
+		DetailsURL: &cra.DetailsURL,
```
Comment (dirgim) [inline git/github/github.go:237]: I think this line should be removed in light of the added logic on the lines below ([current link](https://github.com/redhat-appstudio/integration-service/pull/651/files#diff-9896ed3f1575fe48878f4f5d1b4d39680607d849a5fa3604fe745f0ab79efbb4R256-R260)), that only sets the `detailsURL` if `cra.DetailsURL` is not empty.

File: status/reporter_github.go (hunk 4 above)
Comment (hongweiliu17) [reply status/reporter_github.go]: > so we check if report.Status is in progress then we update the CheckRun and CommitStatus with the Console ITS PLR URL , else we set the **detailsURL** and **targetURL** to **""**

I am not sure if we can set **detailsURL** and **targetURL** to **""** because I don't know what will be shown if that link to **detailsURL** and **targetURL** is clicked. Or don't set it if PLR URL is ""?

Comment (dirgim) [reply status/reporter_github.go]: @hongweiliu17 I think that's what happens with the latest change [for detailsURL](https://github.com/redhat-appstudio/integration-service/pull/651/files#diff-9896ed3f1575fe48878f4f5d1b4d39680607d849a5fa3604fe745f0ab79efbb4R258-R260), not seeing the same for `targetURL`. I'd guess we'd need confirmation in testing on how this would look.

Comment (hongweiliu17) [reply status/reporter_github.go]: > @hongweiliu17 I think that's what happens with the latest change [for detailsURL](https://github.com/redhat-appstudio/integration-service/pull/651/files#diff-9896ed3f1575fe48878f4f5d1b4d39680607d849a5fa3604fe745f0ab79efbb4R258-R260), not seeing the same for `targetURL`. I'd guess we'd need confirmation in testing on how this would look.

I agree.

Comment (kasemAlem) [reply status/reporter_github.go]: for the `targetURL` I can suggest like below 

```
func (c *Client) CreateCommitStatus(ctx context.Context, owner string, repo string, SHA string, state string, description string, statusContext string, targetURL string) (int64, error) {

	var status *ghapi.RepoStatus
	var err error

	if targetURL != "" {
		status, _, err = c.GetRepositoriesService().CreateStatus(ctx, owner, repo, SHA, &ghapi.RepoStatus{State: &state, Description: &description, Context: &statusContext, TargetURL: &targetURL})

	} else {
		status, _, err = c.GetRepositoriesService().CreateStatus(ctx, owner, repo, SHA, &ghapi.RepoStatus{State: &state, Description: &description, Context: &statusContext})
	}

	if err != nil {
		return 0, fmt.Errorf("failed to create an existing commitStatus for GitHub owner/repo/ref %s/%s/%s: %w", owner, repo, SHA, err)
	}

	c.logger.Info("Created commit status",
		"ID", status.ID,
		"Owner", owner,
		"Repository", repo,
		"SHA", SHA,
		"State", state,
		"TargetURL", targetURL,
	)

	return *status.ID, nil
}
```
@dirgim @hongweiliu17 WDYT?

File: git/github/github.go [hunk 6]
```diff
@@ -424,12 +432,12 @@ func (c *Client) GetAllCommentsForPR(ctx context.Context, owner string, repo str
 // CommitStatusExists returns if a match is found for the SHA, state, context and decription.
 func (c *Client) CommitStatusExists(res []*ghapi.RepoStatus, commitStatus *CommitStatusAdapter) (bool, error) {
 	for _, cs := range res {
-		if *cs.State == commitStatus.State && *cs.Description == commitStatus.Description && *cs.Context == commitStatus.Context {
-			c.logger.Info("Found CommitStatus with matching conditions", "CommitStatus.State", commitStatus.State, "CommitStatus.Description", commitStatus.Description, "CommitStatus.Context", commitStatus.Context)
+		if *cs.State == commitStatus.State && *cs.Description == commitStatus.Description && *cs.Context == commitStatus.Context && *cs.TargetURL == commitStatus.TargetURL {
```
[the remark this answers is not in the document]
Comment (kasemAlem) [reply git/github/github.go]: fixed

File: status/reporter_github.go (hunk 4 above)
Comment (dirgim) [reply status/reporter_github.go]: What about something like this (untested):
```
func (c *Client) CreateCommitStatus(ctx context.Context, owner string, repo string, SHA string, state string, description string, statusContext string, targetURL string) (int64, error) {
	repoStatus := ghapi.RepoStatus{}
	repoStatus.State = &state
	repoStatus.Description = &description
	repoStatus.Context = &statusContext

	if targetURL != "" {
		repoStatus.TargetURL = &targetURL
	}

	status, _, err := c.GetRepositoriesService().CreateStatus(ctx, owner, repo, SHA, &repoStatus)
	if err != nil {
		return 0, fmt.Errorf("failed to create an existing commitStatus for GitHub owner/repo/ref %s/%s/%s: %w", owner, repo, SHA, err)
	}

	c.logger.Info("Created commit status",
		"ID", status.ID,
		"Owner", owner,
		"Repository", repo,
		"SHA", SHA,
		"State", state,
		"TargetURL", targetURL,
	)
	return *status.ID, nil
}
```

Comment (kasemAlem) [reply status/reporter_github.go]: thanks @dirgim , it does looks and work better  
I would suggest more improvement on your suggestion : 
 
 instead : 
```
 	repoStatus := ghapi.RepoStatus{}
	repoStatus.State = &state
	repoStatus.Description = &description
	repoStatus.Context = &statusContext
 
```
 to : 
```
	repoStatus := ghapi.RepoStatus{
		State:       &state,
		Description: &description,
		Context:     &statusContext,
	}
```
and to add the **status** to the logger message : 

```
	c.logger.Info("Created commit status",
		"ID", status.ID,
		"Owner", owner,
		"Repository", repo,
		"SHA", SHA,
		"State", state,
		"TargetURL", targetURL,
		"Status", status, // Log the entire status object
```

Comment (dirgim) [reply status/reporter_github.go]: Yup, that works too.

**Review by hongweiliu17 — APPROVED** [review]
LGTM

**Review by dirgim — APPROVED** [review]
LGTM

**Review by MartinBasti — APPROVED** [review]

File: status/status.go [hunk 7]
```diff
@@ -292,15 +292,16 @@ func (s *Status) generateTestReport(ctx context.Context, detail intgteststat.Int
...
+		Text:                text,
+		FullName:            fullName,
+		ScenarioName:        detail.ScenarioName,
+		SnapshotName:        snapshot.Name,
+		ComponentName:       snapshot.Labels[gitops.SnapshotComponentLabel],
+		Status:              detail.Status,
+		Summary:             summary,
+		StartTime:           detail.StartTime,
+		CompletionTime:      detail.CompletionTime,
+		TestPipelineRunName: detail.TestPipelineRunName,
```
Comment (14rcole) [inline status/status.go:304]: It looks like you added this in so that you could call `FormatPipelineURL` in `reporter_github.go`.  Why not just add a URL parameter to the report and call `FormatPipelineURL` here?

I ask because  this is the approach I'm taking in STONEINTG-842.  It works, but whatever we do should be the same so I can switch to your approach if there's some advantage to it.
